### PR TITLE
Make mecha obey set prefix command, make case-insensitive

### PR DESCRIPTION
**Pull request by nathanctech**
Hopefully I'm doing this correctly. :-) ~masterxc

### CODE REVIEW

File: Modules/rat_command.py [hunk 1]
```diff
@@ -85,10 +85,14 @@ class Commands:
...
+            # command didn't start with prefix, ignore it
+        if not message.startswith(cls.prefix):
+            cls.log.debug(f"caught a command without prefix: {message}")
+            return None
         cls.log.debug(f"triggered! message is {message}")
 
-        # remove command prefix
-        raw_command: str = message.lstrip(cls.prefix)
+        # remove command prefix and make lowercase
+        raw_command: str = message.lstrip(cls.prefix).lower()
```
Comment (theunkn0wn1) [inline Modules/rat_command.py]: it amy be advisable to also cast command registrations to lower case, otherwise we can get commands not found that should be found

File: Modules/rat_command.py [hunk 2]
```diff
@@ -85,10 +85,14 @@ class Commands:
             # someone didn't set me.
             raise CommandException(f"Bot client has not been created or not handed to Commands.")
 
+            # command didn't start with prefix, ignore it
+        if not message.startswith(cls.prefix):
+            cls.log.debug(f"caught a command without prefix: {message}")
+            return None
```
Comment (theunkn0wn1) [inline Modules/rat_command.py]: This should probably raise an exception, rather than a null return.

**Comment by theunkn0wn1** [comment]
Alright, well the tests are working now. You now need to write additional tests to increase code coverage to in excess of 85%. right now you are at 71.

File: Modules/rat_command.py [hunk 3]
```diff
@@ -85,10 +92,14 @@ class Commands:
             # someone didn't set me.
             raise CommandException(f"Bot client has not been created or not handed to Commands.")
 
+            # command didn't start with prefix, ignore it
+        if not message.startswith(cls.prefix):
+            cls.log.debug(f"caught a command without prefix: {message}")
+            raise PrefixInvalidException(f"I got an unknown prefix!'")
```
Comment (MHajoha) [inline Modules/rat_command.py]: Hey, good thing you added this check. (I completely forgot to reimplement that after I redid the command invocation)

However, raising an exception on expected behaviour is something that should generally be avoided. 

More importantly, this case will happen tons of times (on each message which isn't a command, likely more often than proper command invocations) and exceptions aren't the most performant thing in programming, especially when they are propagated all the way to the event loop (i.e. not caught by us anywhere) and printed, as would be the case here.

Long story short, this would likely grind Mecha to a snails pace, simply not doing anything / returning immediately would be preferable. (Doing this check in `MechaClient.on_message` would be the best solution in my opinion as it gets rid of an unnecessary function call, but that part is debatable)

Comment (nathanctech) [reply Modules/rat_command.py]: Yeah the initial code was to just return None, but unknown wanted an exception thrown. I preferred the first method although in my testing it only fires on non-alphanumeric characters as the first character like a command trigger (so *, ., + trigger it but normal text does not) so it won't fire as often. I'm fine with either method that makes sense.

I'll correct my git settings, thanks. This is the first ever project I've contributed to so I'm learning still.

Comment (theunkn0wn1) [reply Modules/rat_command.py]: The way i read it is this condition should not be reached, as it shouldn't get past the on_message handler. Therefore this would not be a normal runtime condition and instead an error condition.
That said, the check for the command prefix should probably occur in on_message as that would be the first step in the command execution chain.

**Comment by nathanctech** [comment]
I moved the check to `on_message` since it makes the most sense there. Overhead for a single function call isn't high in the grand scheme of things but no sense processing a command that will get discard anyway.

File: Modules/rat_command.py [hunk 4]
```diff
@@ -55,6 +55,13 @@ class NameCollisionException(CommandException):
     pass
 
 
+class PrefixInvalidException(CommandException):
```
Comment (theunkn0wn1) [inline Modules/rat_command.py]: Is this still necessary? If someone uses the wrong prefix we should ignore it akin any other message coming from the pipe

File: Modules/rat_command.py [hunk 5]
```diff
@@ -129,6 +136,9 @@ def _register(cls, func, names: list or str) -> bool:
         if isinstance(names, str):
             names = [names]  # idiot proofing
 
+        # transform commands to lowercase
+        names = [x.lower() for x in names]
```
Comment (theunkn0wn1) [inline Modules/rat_command.py]: Usage of a single char variable. Given the context, im going to have to ask you to change this.

File: main.py [hunk 6]
```diff
@@ -86,6 +86,11 @@ class MechaClient(Client):
             log.debug("received message from myself ignoring!.")
             return None
 
+        if not str(message).startswith(Commands.prefix):
+            # prevent bot from processing commands without the set prefix
+            log.debug(f"caught a command without prefix. Ignoring.")
```
Comment (theunkn0wn1) [inline main.py]: How do we know it is a command, specifically if it does not have the prefix?
May i suggest rewording it to something more along the lines of 
```py 
log.debug(f"ignoring message {message} as it does not start with my prefix")
```

File: main.py [hunk 7]
```diff
@@ -86,6 +86,11 @@ class MechaClient(Client):
...
+            return None
```
Comment (MHajoha) [inline main.py]: Hey, thanks for checking this here now. Raising a special exception in the trigger method isn't necessary any longer, as unknown pointed out. However a small `assert message.startswith(thingy)` in the trigger method wouldn't hurt and be quite cheap. Assertions could even be disabled for production.

One last thing: while it does help with completion in ides, there's no need to explicitly call `str(message)` here, it would only copy the message string, which we don't need.
Python3(.?) does have a syntax for type hinting, much like statically typed languages. In the arguments, `message: str` will enable the ide of your choice (pycharm?) to provide code completion without a runtime hit to performance.

Comment (nathanctech) [reply main.py]: Made the changes, thanks for the tips. An assert will definitely point out any funny business with that bit of code.

File: main.py (hunk 6 above)
Comment (nathanctech) [reply main.py]: Updated.

File: Modules/rat_command.py (hunk 5 above)
Comment (nathanctech) [reply Modules/rat_command.py]: Also updated, new variable is lowered_name.

File: Modules/rat_command.py [hunk 8]
```diff
@@ -129,6 +132,9 @@ def _register(cls, func, names: list or str) -> bool:
         if isinstance(names, str):
             names = [names]  # idiot proofing
 
+        # transform commands to lowercase
+        names = [lowered_name.lower() for lowered_name in names]
```
Comment (MHajoha) [inline Modules/rat_command.py]: Hey, this is a really tiny nitpick and also the last one I could find, but since you're lowering it only now, `lowered_name` seems unfitting.

Comment (theunkn0wn1) [reply Modules/rat_command.py]: methinks ```py
names = [name.lower() for name in names
```
would be more appropriate

Comment (MHajoha) [reply Modules/rat_command.py]: Yeah
